README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MySQL components for Pip.Service in Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module contains the following packages:
 
- **Build** - a standard factory for constructing components
- **Connect** - instruments for configuring connections to the database.
- **Persistence** - abstract classes for working with the database that can be used for connecting to collections and performing basic CRUD operations

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-mysql-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-mysql-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**, **Dmitrii Uzdemir** and **Dmitrii Levichev**.

The documentation is written by **Mark Makarychev**.

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MySQL components for Pip.Services in Golang / ES2017 Changelog

## <a name="1.0.1"></a> 1.0.1 (2022-10-18) 

Initial public release

### Features
* Added DefaultMySqlFactory
* Added MySqlConnectionResolver
* Added IdentifiableJsonMySqlPersistence
* Added IdentifiableMySqlPersistence
* Added MySqlConnection
* Added MySqlPersistence





test/connect/MysqlConnection_test.go
package test_connect

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	conn "github.com/pip-services3-gox/pip-services3-mysql-gox/connect"
	"github.com/stretchr/testify/assert"
)

func TestMySqlConnection(t *testing.T) {
	var connection *conn.MySqlConnection

	mysqlUri := os.Getenv("MYSQL_URI")
	mysqlHost := os.Getenv("MYSQL_HOST")
	if mysqlHost == "" {
		mysqlHost = "localhost"
	}
	mysqlPort := os.Getenv("MYSQL_PORT")
	if mysqlPort == "" {
		mysqlPort = "3306"
	}
	mysqlDatabase := os.Getenv("MYSQL_DB")
	if mysqlDatabase == "" {
		mysqlDatabase = "test"
	}
	mysqlUser := os.Getenv("MYSQL_USER")
	if mysqlUser == "" {
		mysqlUser = "mysql"
	}
	mysqlPassword := os.Getenv("MYSQL_PASSWORD")
	if mysqlPassword == "" {
		mysqlPassword = "mysql"
	}

	if mysqlUri == "" && mysqlHost == "" {
		t.Skip("No MySql credentials")
	}

	dbConfig := cconf.NewConfigParamsFromTuples(
		"connection.uri", mysqlUri,
		"connection.host", mysqlHost,
		"connection.port", mysqlPort,
		"connection.database", mysqlDatabase,
		"credential.username", mysqlUser,
		"credential.password", mysqlPassword,
		"options.max_pool_size", 10,
		"options.connect_timeout", 100,
		"options.idle_timeout", 100,
	)

	connection = conn.NewMySqlConnection()
	connection.Configure(context.Background(), dbConfig)
	err := connection.Open(context.Background(), "")
	assert.Nil(t, err)

	assert.NotNil(t, connection.GetConnection())
	assert.NotEmpty(t, connection.GetDatabaseName())
	assert.NotNil(t, connection.GetDatabaseName())

	err = connection.Close(context.Background(), "")
	assert.Nil(t, err)
}


test/fixtures/DummyPersistenceFixture.go
package fixtures

import (
	"context"
	"testing"

	cdata "github.com/pip-services3-gox/pip-services3-commons-gox/data"
	"github.com/stretchr/testify/assert"
)

type DummyPersistenceFixture struct {
	dummy1      Dummy
	dummy2      Dummy
	persistence IDummyPersistence
}

func NewDummyPersistenceFixture(persistence IDummyPersistence) *DummyPersistenceFixture {
	c := DummyPersistenceFixture{}
	c.dummy1 = Dummy{Id: "", Key: "Key 11", Content: "Content 1"}
	c.dummy2 = Dummy{Id: "", Key: "Key 2", Content: "Content 2"}
	c.persistence = persistence
	return &c
}

func (c *DummyPersistenceFixture) TestCrudOperations(t *testing.T) {
	var dummy1 Dummy
	var dummy2 Dummy

	result, err := c.persistence.Create(context.Background(), "", c.dummy1)
	assert.Nil(t, err)

	dummy1 = result
	assert.NotEqual(t, Dummy{}, dummy1)
	assert.NotNil(t, dummy1.Id)
	assert.Equal(t, c.dummy1.Key, dummy1.Key)
	assert.Equal(t, c.dummy1.Content, dummy1.Content)

	// Create another dummy by send pointer
	result, err = c.persistence.Create(context.Background(), "", c.dummy2)
	assert.Nil(t, err)

	dummy2 = result
	assert.NotEqual(t, Dummy{}, dummy2)
	assert.NotNil(t, dummy2.Id)
	assert.Equal(t, c.dummy2.Key, dummy2.Key)
	assert.Equal(t, c.dummy2.Content, dummy2.Content)

	page, err := c.persistence.GetPageByFilter(context.Background(), "", *cdata.NewEmptyFilterParams(), *cdata.NewPagingParams(0, 5, true))
	assert.Nil(t, err)

	assert.True(t, page.HasData())
	assert.Len(t, page.Data, 2)
	assert.True(t, page.HasTotal())
	assert.Equal(t, page.Total, 2)

	assert.True(t, page.Data[0].Key == dummy1.Key || page.Data[0].Key == dummy2.Key)
	assert.True(t, page.Data[1].Key == dummy1.Key || page.Data[1].Key == dummy2.Key)

	page, err = c.persistence.GetPageByFilter(context.Background(), "",
		*cdata.NewFilterParamsFromTuples("Key", "Key 11"),
		*cdata.NewPagingParams(0, 5, true),
	)
	assert.Nil(t, err)

	assert.True(t, page.HasData())
	assert.Len(t, page.Data, 1)
	assert.True(t, page.HasTotal())
	assert.Equal(t, page.Total, 1)

	assert.Equal(t, page.Data[0].Key, dummy1.Key)

	// Update the dummy
	dummy1.Content = "Updated Content 1"
	result, err = c.persistence.Update(context.Background(), "", dummy1)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy1.Id, result.Id)
	assert.Equal(t, dummy1.Key, result.Key)
	assert.Equal(t, dummy1.Content, result.Content)

	// Set the dummy (updating)
	dummy1.Content = "Updated Content 2"
	result, err = c.persistence.Set(context.Background(), "", dummy1)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy1.Id, result.Id)
	assert.Equal(t, dummy1.Key, result.Key)
	assert.Equal(t, dummy1.Content, result.Content)

	// Set the dummy (creating)
	dummy2.Id = "New_id"
	dummy2.Key = "New_key"
	result, err = c.persistence.Set(context.Background(), "", dummy2)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy2.Id, result.Id)
	assert.Equal(t, dummy2.Key, result.Key)
	assert.Equal(t, dummy2.Content, result.Content)

	// Partially update the dummy
	updateMap := cdata.NewAnyValueMapFromTuples("content", "Partially Updated Content 1")
	result, err = c.persistence.UpdatePartially(context.Background(), "", dummy1.Id, *updateMap)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy1.Id, result.Id)
	assert.Equal(t, dummy1.Key, result.Key)
	assert.Equal(t, "Partially Updated Content 1", result.Content)

	// Get the dummy by Id
	result, err = c.persistence.GetOneById(context.Background(), "", dummy1.Id)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy1.Id, result.Id)
	assert.Equal(t, dummy1.Key, result.Key)
	assert.Equal(t, "Partially Updated Content 1", result.Content)

	// Delete the dummy
	result, err = c.persistence.DeleteById(context.Background(), "", dummy1.Id)
	assert.Nil(t, err)

	assert.NotEqual(t, Dummy{}, result)
	assert.Equal(t, dummy1.Id, result.Id)
	assert.Equal(t, dummy1.Key, result.Key)
	assert.Equal(t, "Partially Updated Content 1", result.Content)

	// Get the deleted dummy
	result, err = c.persistence.GetOneById(context.Background(), "", dummy1.Id)
	assert.Nil(t, err)
	assert.Equal(t, Dummy{}, result)
}

func (c *DummyPersistenceFixture) TestBatchOperations(t *testing.T) {
	var dummy1 Dummy
	var dummy2 Dummy

	// Create one dummy
	result, err := c.persistence.Create(context.Background(), "", c.dummy1)
	assert.Nil(t, err)

	dummy1 = result
	assert.NotEqual(t, Dummy{}, dummy1)
	assert.NotNil(t, dummy1.Id)
	assert.Equal(t, c.dummy1.Key, dummy1.Key)
	assert.Equal(t, c.dummy1.Content, dummy1.Content)

	// Create another dummy
	result, err = c.persistence.Create(context.Background(), "", c.dummy2)
	assert.Nil(t, err)

	dummy2 = result
	assert.NotEqual(t, Dummy{}, dummy2)
	assert.NotNil(t, dummy2.Id)
	assert.Equal(t, c.dummy2.Key, dummy2.Key)
	assert.Equal(t, c.dummy2.Content, dummy2.Content)

	// Read batch
	items, err := c.persistence.GetListByIds(context.Background(), "", []string{dummy1.Id, dummy2.Id})
	assert.Nil(t, err)

	assert.NotNil(t, items)
	assert.Len(t, items, 2)

	// Delete batch
	err = c.persistence.DeleteByIds(context.Background(), "", []string{dummy1.Id, dummy2.Id})
	assert.Nil(t, err)

	// Read empty batch
	items, err = c.persistence.GetListByIds(context.Background(), "", []string{dummy1.Id, dummy2.Id})
	assert.Nil(t, err)

	assert.Len(t, items, 0)
}

func (c *DummyPersistenceFixture) TestRandomOperation(t *testing.T) {
	var dummy1 Dummy
	var dummy2 Dummy

	result, err := c.persistence.GetOneRandom(context.Background(), "")
	assert.Nil(t, err)
	assert.Equal(t, Dummy{}, result)
	assert.Equal(t, result.Id, "")
	assert.Equal(t, result.Key, "")
	assert.Equal(t, result.Content, "")

	// Create one dummy
	result, err = c.persistence.Create(context.Background(), "", c.dummy1)
	assert.Nil(t, err)

	dummy1 = result
	assert.NotEqual(t, Dummy{}, dummy1)
	assert.Equal(t, c.dummy1.Key, dummy1.Key)
	assert.Equal(t, c.dummy1.Content, dummy1.Content)

	// Create another dummy
	result, err = c.persistence.Create(context.Background(), "", c.dummy2)
	assert.Nil(t, err)

	dummy2 = result
	assert.NotEqual(t, Dummy{}, dummy2)
	assert.Equal(t, c.dummy2.Key, dummy2.Key)
	assert.Equal(t, c.dummy2.Content, dummy2.Content)

	result, err = c.persistence.GetOneRandom(context.Background(), "")
	assert.Nil(t, err)
	assert.NotEqual(t, Dummy{}, result)
}


test/persistence/DummyMysqlPersistence.go
package test

import (
	"context"

	cdata "github.com/pip-services3-gox/pip-services3-commons-gox/data"
	persist "github.com/pip-services3-gox/pip-services3-mysql-gox/persistence"
	"github.com/pip-services3-gox/pip-services3-mysql-gox/test/fixtures"
)

type DummyMySqlPersistence struct {
	*persist.IdentifiableMySqlPersistence[fixtures.Dummy, string]
}

func NewDummyMySqlPersistence() *DummyMySqlPersistence {
	c := &DummyMySqlPersistence{}
	c.IdentifiableMySqlPersistence = persist.InheritIdentifiableMySqlPersistence[fixtures.Dummy, string](c, "dummies")
	return c
}

func (c *DummyMySqlPersistence) DefineSchema() {
	c.ClearSchema()
	c.IdentifiableMySqlPersistence.DefineSchema()
	// Row name must be in double quotes for properly case!!!
	c.EnsureSchema("CREATE TABLE `" + c.TableName + "` (id VARCHAR(32) PRIMARY KEY, `key` VARCHAR(50), `content` TEXT)")
	c.EnsureIndex(c.IdentifiableMySqlPersistence.TableName+"_key", map[string]string{"key": "1"}, map[string]string{"unique": "true"})
}

func (c *DummyMySqlPersistence) GetPageByFilter(ctx context.Context, correlationId string,
	filter cdata.FilterParams, paging cdata.PagingParams) (page cdata.DataPage[fixtures.Dummy], err error) {

	key, ok := filter.GetAsNullableString("Key")
	filterObj := ""
	if ok && key != "" {
		filterObj += "`key`='" + key + "'"
	}
	sorting := ""

	return c.IdentifiableMySqlPersistence.GetPageByFilter(ctx, correlationId,
		filterObj, paging,
		sorting, "",
	)
}

func (c *DummyMySqlPersistence) GetCountByFilter(ctx context.Context, correlationId string,
	filter cdata.FilterParams) (count int64, err error) {

	key, ok := filter.GetAsNullableString("Key")
	filterObj := ""
	if ok && key != "" {
		filterObj += "`key`='" + key + "'"
	}
	return c.IdentifiableMySqlPersistence.GetCountByFilter(ctx, correlationId, filterObj)
}

func (c *DummyMySqlPersistence) GetOneRandom(ctx context.Context, correlationId string) (item fixtures.Dummy, err error) {
	return c.IdentifiableMySqlPersistence.GetOneRandom(ctx, correlationId, "")
}


connect/MysqlConnection.go
package connect

import (
	"context"
	"database/sql"
	"math"
	"time"

	_ "github.com/go-sql-driver/mysql"

	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	clog "github.com/pip-services3-gox/pip-services3-components-gox/log"
)

// MySQL connection using plain driver.
//
// By defining a connection and sharing it through multiple persistence components
// you can reduce number of used database connections.
//
//	Configuration parameters
//		- connection(s):
//			- discovery_key:        (optional) a key to retrieve the connection from IDiscovery
//			- host:                 host name or IP address
//			- port:                 port number (default: 27017)
//			- uri:                  resource URI or connection string with all parameters in it
//		- credential(s):
//			- store_key:            (optional) a key to retrieve the credentials from ICredentialStore
//			- username:             user name
//			- password:             user password
//		- options:
//			- connect_timeout:      (optional) number of milliseconds to wait before timing out when connecting a new client (default: 0)
//			- idle_timeout:         (optional) number of milliseconds a client must sit idle in the pool and not be checked out (default: 10000)
//			- max_pool_size:        (optional) maximum number of clients the pool should contain (default: 10)
//
//	References
//		- *:logger:*:*:1.0           (optional) ILogger components to pass log messages
//		- *:discovery:*:*:1.0        (optional) IDiscovery services
//		- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credentials
type MySqlConnection struct {
	defaultConfig *cconf.ConfigParams
	// The logger.
	Logger *clog.CompositeLogger
	// The connection resolver.
	ConnectionResolver *MySqlConnectionResolver
	// The configuration options.
	Options *cconf.ConfigParams
	// The MySQL connection pool object.
	Connection *sql.DB
	// The MySQL database name.
	DatabaseName string

	retries int
}

const (
	DefaultConnectTimeout = 1000
	DefaultIdleTimeout    = 10000
	DefaultMaxPoolSize    = 3
	DefaultRetriesCount   = 3
)

// NewMySqlConnection creates a new instance of the connection component.
func NewMySqlConnection() *MySqlConnection {
	c := &MySqlConnection{
		defaultConfig: cconf.NewConfigParamsFromTuples(
			"options.connect_timeout", DefaultConnectTimeout,
			"options.idle_timeout", DefaultIdleTimeout,
			"options.max_pool_size", DefaultMaxPoolSize,
		),
		Logger:             clog.NewCompositeLogger(),
		ConnectionResolver: NewMySqlConnectionResolver(),
		Options:            cconf.NewEmptyConfigParams(),
		retries:            DefaultRetriesCount,
	}
	return c
}

// Configure component by passing configuration parameters.
//	Parameters:
//		- ctx context.Context
//		- config configuration parameters to be set.
func (c *MySqlConnection) Configure(ctx context.Context, config *cconf.ConfigParams) {
	config = config.SetDefaults(c.defaultConfig)
	c.ConnectionResolver.Configure(ctx, config)
	c.Options = c.Options.Override(config.GetSection("options"))

	c.DatabaseName, _ = config.GetAsNullableString("connection.database")
}

// SetReferences references to dependent components.
//	Parameters:
//		- ctx context.Context
//		- references references to locate the component dependencies.
func (c *MySqlConnection) SetReferences(ctx context.Context, references cref.IReferences) {
	c.Logger.SetReferences(ctx, references)
	c.ConnectionResolver.SetReferences(ctx, references)
}

// IsOpen checks if the component is opened.
//	Returns true if the component has been opened and false otherwise.
func (c *MySqlConnection) IsOpen() bool {
	return c.Connection != nil
}

//	Open the component.
//	Parameters:
//		- ctx context.Context
//		- correlationId 	(optional) transaction id to trace execution through call chain.
//		- Return 			error or nil no errors occurred.
func (c *MySqlConnection) Open(ctx context.Context, correlationId string) error {

	uri, err := c.ConnectionResolver.Resolve(ctx, correlationId)
	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "Failed to resolve MySql connection")
		return nil
	}

	c.Logger.Debug(ctx, correlationId, "Connecting to mysql")

	retries := c.retries
	for retries > 0 {
		pool, err := sql.Open("mysql", uri)
		if err != nil {
			retries--
			if retries <= 0 {
				return cerr.
					NewConnectionError(correlationId, "CONNECT_FAILED", "Connection to mysql failed").
					WithCause(err)
			}
			c.Logger.Debug(ctx, correlationId, "Failed to connect to mysqls, try reconnect...")
			err = c.waitForRetry(ctx, correlationId, retries)
			if err != nil {
				return err
			}
			continue
		}
		idleTimeoutMS := c.Options.GetAsIntegerWithDefault("idle_timeout", DefaultIdleTimeout)
		maxPoolSize := c.Options.GetAsIntegerWithDefault("max_pool_size", DefaultMaxPoolSize)
		connectTimeoutMS := c.Options.GetAsIntegerWithDefault("connect_timeout", DefaultConnectTimeout)

		pool.SetConnMaxIdleTime(time.Duration(idleTimeoutMS) * time.Millisecond)
		pool.SetMaxOpenConns(maxPoolSize)
		pool.SetConnMaxLifetime(time.Duration(connectTimeoutMS) * time.Millisecond)

		c.Connection = pool
		break
	}
	return nil
}

// Close component and frees used resources.
//	Parameters:
//		- ctx context.Context
//		- correlationId (optional) transaction id to trace execution through call chain.
//	Returns: error or nil no errors occurred
func (c *MySqlConnection) Close(ctx context.Context, correlationId string) error {
	if c.Connection == nil {
		return nil
	}
	c.Connection.Close()
	c.Logger.Debug(ctx, correlationId, "Disconnected from mysql database %s", c.DatabaseName)
	c.Connection = nil
	c.DatabaseName = ""
	return nil
}

func (c *MySqlConnection) GetConnection() *sql.DB {
	return c.Connection
}

func (c *MySqlConnection) GetDatabaseName() string {
	return c.DatabaseName
}

func (c *MySqlConnection) waitForRetry(ctx context.Context, correlationId string, retries int) error {
	waitTime := DefaultConnectTimeout * int(math.Pow(float64(c.retries-retries), 2))

	select {
	case <-time.After(time.Duration(waitTime) * time.Millisecond):
		return nil
	case <-ctx.Done():
		return cerr.ApplicationErrorFactory.Create(
			&cerr.ErrorDescription{
				Type:          "Application",
				Category:      "Application",
				Code:          "CONTEXT_CANCELLED",
				Message:       "request canceled by parent context",
				CorrelationId: correlationId,
			},
		)
	}
}


persistence/IdentifiableMysqlPersistence.go
package persistence

import (
	"context"

	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cpersist "github.com/pip-services3-gox/pip-services3-data-gox/persistence"

	cdata "github.com/pip-services3-gox/pip-services3-commons-gox/data"
)

// IdentifiableMySqlPersistence Abstract persistence component that stores data in MySQL
// and implements a number of CRUD operations over data items with unique ids.
// The data items must implement IIdentifiable interface.
//
// In basic scenarios child classes shall only override getPageByFilter,
// getListByFilter or deleteByFilter operations with specific filter function.
// All other operations can be used out of the box.
//
// In complex scenarios child classes can implement additional operations by
// accessing c._collection and c._model properties.
//
//	Configuration parameters
//		- collection:               (optional) MySQL collection name
//		- connection(s):
//			- discovery_key:        (optional) a key to retrieve the connection from IDiscovery
//			- host:                 host name or IP address
//			- port:                 port number (default: 27017)
//			- uri:                  resource URI or connection string with all parameters in it
//		- credential(s):
//			- store_key:            (optional) a key to retrieve the credentials from ICredentialStore
//			- username:             (optional) user name
//			- password:             (optional) user password
//		- options:
//			- connect_timeout:      (optional) number of milliseconds to wait before timing out when connecting a new client (default: 0)
//			- idle_timeout:         (optional) number of milliseconds a client must sit idle in the pool and not be checked out (default: 10000)
//			- max_pool_size:        (optional) maximum number of clients the pool should contain (default: 10)
//
//	References
//		- *:logger:*:*:1.0           (optional) ILogger components to pass log messages components to pass log messages
//		- *:discovery:*:*:1.0        (optional) IDiscovery services
//		- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credentials
//
// Example:
//	type MyMySqlPersistence struct {
//		*persist.IdentifiableMySqlPersistence[MyData, string]
//	}
//
//	func NewMyMySqlPersistence() *MyMySqlPersistence {
//		c := &MyMySqlPersistence{}
//		c.IdentifiableMySqlPersistence = persist.InheritIdentifiableMySqlPersistence[MyData, string](c, "mydata")
//		return c
//	}
//
//	func (c *MyMySqlPersistence) DefineSchema() {
//		c.ClearSchema()
//		c.IdentifiableMySqlPersistence.DefineSchema()
//		// Row name must be in double quotes for properly case!!!
//		c.EnsureSchema("CREATE TABLE `" + c.TableName + "` (id VARCHAR(32) PRIMARY KEY, `key` VARCHAR(50), `content` TEXT)")
//		c.EnsureIndex(c.IdentifiableMySqlPersistence.TableName+"_key", map[string]string{"key": "1"}, map[string]string{"unique": "true"})
//	}
//
//	func (c *MyMySqlPersistence) GetPageByFilter(ctx context.Context, correlationId string,
//		filter cdata.FilterParams, paging cdata.PagingParams) (page cdata.DataPage[MyData], err error) {
//
//		key, ok := filter.GetAsNullableString("Key")
//		filterObj := ""
//		if ok && key != "" {
//			filterObj += "`key`='" + key + "'"
//		}
//		sorting := ""
//
//		return c.IdentifiableMySqlPersistence.GetPageByFilter(ctx, correlationId,
//			filterObj, paging,
//			sorting, "",
//		)
//	}
//
//	func main() {
//		persistence := NewMyMySqlPersistence()
//
//		persistence.Configure(context.Background(), NewConfigParamsFromTuples(
//			"host", "localhost",
//			"port", 27017,
//		))
//
//		err := persistence.Open(context.Background(), "123")
//
//		item, err := persistence.Create(context.Background(), "123", MyData{Id: "1", Name: "ABC"})
//		page, err := persistence.GetPageByFilter(context.Background(), "123", *NewFilterParamsFromTuples("name", "ABC"), nil)
//
//		fmt.Println(page.Data)
//		res, err := persistence.DeleteById(context.Background(), "123", "1")
//	}
//
type IdentifiableMySqlPersistence[T any, K any] struct {
	*MySqlPersistence[T]
}

// InheritIdentifiableMySqlPersistence creates a new instance of the persistence component.
//	Parameters:
//		- ctx context.Context
//		- overrides References to override virtual methods
//		- tableName    (optional) a table name.
func InheritIdentifiableMySqlPersistence[T any, K any](overrides IMySqlPersistenceOverrides[T], tableName string) *IdentifiableMySqlPersistence[T, K] {
	if tableName == "" {
		panic("Table name could not be empty")
	}

	c := &IdentifiableMySqlPersistence[T, K]{}
	c.MySqlPersistence = InheritMySqlPersistence[T](overrides, tableName)

	return c
}

// GetListByIds gets a list of data items retrieved by given unique ids.
//	Parameters:
//		- ctx context.Context
//		- correlationId     (optional) transaction id to trace execution through call chain.
//		- ids of data items to be retrieved
//	Returns: a data list or error.
func (c *IdentifiableMySqlPersistence[T, K]) GetListByIds(ctx context.Context, correlationId string,
	ids []K) (items []T, err error) {

	ln := len(ids)
	params := c.GenerateParameters(ln)
	query := "SELECT * FROM " + c.QuotedTableName() + " WHERE id IN(" + params + ")"

	rows, err := c.Client.QueryContext(ctx, query, ItemsToAnySlice(ids)...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	items = make([]T, 0)
	for rows.Next() {
		if c.IsTerminated() {
			rows.Close()
			return nil, cerr.
				NewError("query terminated").
				WithCorrelationId(correlationId)
		}
		item, convErr := c.Overrides.ConvertToPublic(rows)
		if convErr != nil {
			return items, convErr
		}
		items = append(items, item)
	}

	if items != nil {
		c.Logger.Trace(ctx, correlationId, "Retrieved %d from %s", len(items), c.TableName)
	}

	return items, rows.Err()
}

// GetOneById gets a data item by its unique id.
//	Parameters:
//		- ctx context.Context
//		- correlationId     (optional) transaction id to trace execution through call chain.
//		- id                an id of data item to be retrieved.
// Returns: data item or error.
func (c *IdentifiableMySqlPersistence[T, K]) GetOneById(ctx context.Context, correlationId string, id K) (item T, err error) {

	query := "SELECT * FROM " + c.QuotedTableName() + " WHERE id=?"

	rows, err := c.Client.QueryContext(ctx, query, id)
	if err != nil {
		return item, err
	}
	defer rows.Close()

	if !rows.Next() {
		return item, rows.Err()
	}

	if err == nil {
		c.Logger.Trace(ctx, correlationId, "Retrieved from %s with id = %s", c.TableName, id)
		return c.Overrides.ConvertToPublic(rows)
	}
	c.Logger.Trace(ctx, correlationId, "Nothing found from %s with id = %s", c.TableName, id)
	return item, err
}

// Create a data item.
//	Parameters:
//		- ctx context.Context
//		- correlation_id    (optional) transaction id to trace execution through call chain.
//		- item              an item to be created.
//	Returns: (optional)  created item or error.
func (c *IdentifiableMySqlPersistence[T, K]) Create(ctx context.Context, correlationId string, item T) (result T, err error) {
	newItem := c.cloneItem(item)
	newItem = GenerateObjectIdIfNotExists[T](newItem)

	return c.MySqlPersistence.Create(ctx, correlationId, newItem)
}

// Set a data item. If the data item exists it updates it,
// otherwise it creates a new data item.
//	Parameters:
//		- ctx context.Context
//		- correlation_id    (optional) transaction id to trace execution through call chain.
//		- item              an item to be set.
//	Returns: (optional)  updated item or error.
func (c *IdentifiableMySqlPersistence[T, K]) Set(ctx context.Context, correlationId string, item T) (result T, err error) {
	objMap, convErr := c.Overrides.ConvertFromPublic(item)
	if convErr != nil {
		return result, convErr
	}

	GenerateObjectMapIdIfNotExists(objMap)

	columns, values := c.GenerateColumnsAndValues(objMap)

	paramsStr := c.GenerateParameters(len(values))
	columnsStr := c.GenerateColumns(columns)
	setParams := c.GenerateSetParameters(columns)
	id := cpersist.GetObjectId(objMap)

	values = append(values, values...)

	query := "INSERT INTO " + c.QuotedTableName() + " (" + columnsStr + ") VALUES (" + paramsStr + ")"
	query += " ON DUPLICATE KEY UPDATE " + setParams

	_, err = c.Client.ExecContext(ctx, query, values...)
	if err != nil {
		return result, err
	}

	// Getting result
	query = "SELECT * FROM " + c.QuotedTableName() + " WHERE id=?"
	rows, err := c.Client.QueryContext(ctx, query, []any{id}...)
	if err != nil {
		return result, err
	}
	defer rows.Close()

	if !rows.Next() {
		return result, rows.Err()
	}

	if err == nil {
		result, convErr = c.Overrides.ConvertToPublic(rows)
		if convErr != nil {
			return result, convErr
		}
		c.Logger.Trace(ctx, correlationId, "Set in %s with id = %s", c.TableName, id)
		return result, nil
	}
	return result, rows.Err()

}

// Update a data item.
//	Parameters:
//		- ctx context.Context
//		- correlation_id    (optional) transaction id to trace execution through call chain.
//		- item              an item to be updated.
//	Returns          (optional)  updated item or error.
func (c *IdentifiableMySqlPersistence[T, K]) Update(ctx context.Context, correlationId string, item T) (result T, err error) {
	objMap, convErr := c.Overrides.ConvertFromPublic(item)
	if convErr != nil {
		return result, convErr
	}
	columns, values := c.GenerateColumnsAndValues(objMap)
	paramsStr := c.GenerateSetParameters(columns)
	id := cpersist.GetObjectId(objMap)
	values = append(values, id)

	query := "UPDATE " + c.QuotedTableName() + " SET " + paramsStr + " WHERE id=?"

	_, err = c.Client.ExecContext(ctx, query, values...)
	if err != nil {
		return result, err
	}

	// Getting result
	query = "SELECT * FROM " + c.QuotedTableName() + " WHERE id=?"
	rows, err := c.Client.QueryContext(ctx, query, []any{id}...)
	if err != nil {
		return result, err
	}

	defer rows.Close()
	if !rows.Next() {
		return result, rows.Err()
	}

	if err == nil {
		result, convErr = c.Overrides.ConvertToPublic(rows)
		if convErr != nil {
			return result, convErr
		}
		c.Logger.Trace(ctx, correlationId, "Updated in %s with id = %s", c.TableName, id)
		return result, nil
	}
	return result, err
}

// UpdatePartially updates only few selected fields in a data item.
//	Parameters:
//		- ctx context.Context
//		- correlation_id    (optional) transaction id to trace execution through call chain.
//		- id                an id of data item to be updated.
//		- data              a map with fields to be updated.
//	Returns: updated item or error.
func (c *IdentifiableMySqlPersistence[T, K]) UpdatePartially(ctx context.Context, correlationId string, id K, data cdata.AnyValueMap) (result T, err error) {
	objMap, convErr := c.Overrides.ConvertFromPublicPartial(data.Value())
	if convErr != nil {
		return result, convErr
	}
	columns, values := c.GenerateColumnsAndValues(objMap)
	paramsStr := c.GenerateSetParameters(columns)
	values = append(values, id)

	query := "UPDATE " + c.QuotedTableName() + " SET " + paramsStr + " WHERE id=?"

	_, err = c.Client.ExecContext(ctx, query, values...)
	if err != nil {
		return result, err
	}

	query = "SELECT * FROM " + c.QuotedTableName() + " WHERE id=?"
	rows, err := c.Client.QueryContext(ctx, query, []any{id}...)
	if err != nil {
		return result, err
	}
	defer rows.Close()

	if !rows.Next() {
		return result, rows.Err()
	}

	if err == nil {
		result, convErr = c.Overrides.ConvertToPublic(rows)
		if convErr != nil {
			return result, convErr
		}
		c.Logger.Trace(ctx, correlationId, "Updated partially in %s with id = %s", c.TableName, id)
		return result, nil
	}
	return result, rows.Err()
}

// DeleteById deletes a data item by its unique id.
//	Parameters:
//		- ctx context.Context
//		- correlation_id    (optional) transaction id to trace execution through call chain.
//		- id                an id of the item to be deleted
//	Returns: (optional)  deleted item or error.
func (c *IdentifiableMySqlPersistence[T, K]) DeleteById(ctx context.Context, correlationId string, id K) (result T, err error) {
	query := "SELECT * FROM " + c.QuotedTableName() + " WHERE id=?"

	rows, err := c.Client.QueryContext(ctx, query, []any{id}...)
	if err != nil {
		return result, err
	}

	query = "DELETE FROM " + c.QuotedTableName() + " WHERE id=?"
	_, err = c.Client.ExecContext(ctx, query, []any{id}...)
	if err != nil {
		return result, err
	}
	defer rows.Close()

	if !rows.Next() {
		return result, rows.Err()
	}

	if err == nil {
		result, convErr := c.Overrides.ConvertToPublic(rows)
		if convErr != nil {
			return result, convErr
		}
		c.Logger.Trace(ctx, correlationId, "Deleted from %s with id = %s", c.TableName, id)
		return result, nil
	}
	return result, rows.Err()
}

// DeleteByIds deletes multiple data items by their unique ids.
//	Parameters:
//		- ctx context.Context
//		- correlationId     (optional) transaction id to trace execution through call chain.
//		- ids                of data items to be deleted.
//	Returns: (optional)  error or null for success.
func (c *IdentifiableMySqlPersistence[T, K]) DeleteByIds(ctx context.Context, correlationId string, ids []K) error {

	ln := len(ids)
	paramsStr := c.GenerateParameters(ln)

	query := "DELETE FROM " + c.QuotedTableName() + " WHERE id IN(" + paramsStr + ")"

	result, err := c.Client.ExecContext(ctx, query, ItemsToAnySlice(ids)...)
	if err != nil {
		return err
	}

	count, err := result.RowsAffected()
	if err != nil {
		return err
	}

	if count != 0 {
		c.Logger.Trace(ctx, correlationId, "Deleted %d items from %s", count, c.TableName)
	}
	return nil
}
